### DISCUSSION TITLE
BatchedExecutor: The only sane executor.

### DISCUSSION THREAD

**Discussion by follyfoxe** [SciSharp/LLamaSharp#804] [General]
### Initial pain
I've experimented with LlamaSharp for a while now, checking if any new features were added every once in a while.
But something has always frustrated me everytime I came back to it: The sheer insanity that is to get the high level executors to work.
It's all so, so messy and almost impossible to get working cleanly.
Things like not being able to directly prompt the models with tokens, or how tokenizing special tokens used to be disabled by default (Which would've been fine if only you were able to pass in your own tokens!)

I work on my laptop, which isn't the most powerful machine out there, so I always test small language models like phi-3 and other models whose parameter sizes are around 7b.
Now, these models being so small, are an absolute pain to control. Often times they dont output the end token, or they begin writing user or system messages.
Meaning the context would get all messy with incorrectly generated tokens and no way to correct it because the default executors didn't allow it.
In fact, you couldn't even use antiprompts to fix that! Because first, the generated token would get added to the cache anyways. (At least that's what I think it did by looking at the source code)
And second, the antiprompts were strings and in the code they checked the string output, so special tokens didnt even work!

### Updates
Now, some of these issues were slowly improved... But not completely solved.
For example special tokens are now included on the high level executors. Allowing for proper prompt formats like chatml.
And the StreamingTokenDecoder now allows displaying special tokens which is also nice, since before we had to resort to the NativeApi class...
I was devastated.

### BatchedExecutor to the rescue
Recently I've decided to give another shot to LLamaSharp, more specifically the new BatchedExecutor which caught my attention.
I looked at the examples and everything slowly clicked into place. I read the source code, and the api was so much clearer.
I proceeded to refactor an old project of mine that used LLamaSharp, and in a matter of minutes, I had phi-3 under control and working as intended and in a much more reliable way than ever before.
To say the least, this thing was incredible.

I really like the fact that I actually got control over the model's output, instead of leaving it to the old executors' messy implementation.
I also like that it doesn't come with some weird state handling shenanigans, that way I can keep track of generated tokens *manually*. I even came up with a way to handle the context running out, take a look!
```
// For context: This is inside a custom wrapping class.
// _tokenHistory and _tokenPrompt are both List<LLamaToken>
// _tokenPrompt is modified by various public methods, kind of like the System.Text.StringBuilder api
// _tokensKeep is an int and is set by the user.
void Prompt()
{
    // Add BOS if needed
    if (_tokenHistory.Count <= 0)
        _tokenPrompt.Insert(0, _specialTokens[SpecialToken.BOS]);
    
    // Make sure the prompt doesn't exceed the context size
    int max = (int)Context.ContextSize;
    if (_tokenPrompt.Count > max)
    {
        _tokenPrompt.RemoveRange(0, _tokenPrompt.Count - max);
        Log("Prompt was too big! Consider reducing it...");
    }
    
    // Make sure theres enough space for the prompt
    int target = _tokenHistory.Count + _tokenPrompt.Count;
    if (target > max)
    {
        int remove = target - max;
        _tokenHistory.RemoveRange(_tokensKeep, remove);
        Conversation.ShiftLeft(remove, _tokensKeep); // Really love that this method exists by default.
        Log("Making space...");
    }

    Conversation.Prompt(_tokenPrompt); // We prompt the main conversation
    _tokenHistory.AddRange(_tokenPrompt);
    _tokenPrompt.Clear();
}
```
I haven't tested it throughly yet, but it just *feels* right.

Absolutely love the fact that to continue generation, you have to manually prompt the executor with the last generated token. This allows me to actually control the stuff that gets sent to the model. Meaning I can stop the generation when I encounter the model trying to generate user messages when it shouldn't, and the context won't get messy!
This is how it should've been from the start.

My inference loop looks something like this, it's not the best, but it's way cleaner than before.
```
// For context, these get set to a Dictionary<SpecialToken, LLamaToken> when the model loads.
public enum SpecialToken
{
    BOS,
    System,
    Assistant,
    User,
    EndMessage
}

// Inference
using var sampler = new DefaultSamplingPipeline();
var decoder = new StreamingTokenDecoder(Context) { DecodeSpecialTokens = true };

Prompt();
for (var i = 0; i < MaxTokens; i++)
{
    await Executor.Infer();
    LLamaToken token = sampler.Sample(Context.NativeHandle, Conversation.Sample(), _tokenHistory);
    if (ModelWeights.Tokens.IsEndOfGeneration(token) || IsSpecialToken(token))
        break;

    Write(token);
    Prompt();
    decoder.Add(token);
    
    DoStuff(decoder.Read());
}

Write(SpecialToken.EndMessage); // The next inference will include this.
WriteLine();
```

These code snippets were taken out of context so don't expect them to work by just copy pasting; They're just for demonstration.

### Where are we now
In a more serious note, I decided to make this with the purpose of giving feedback to the project and hopefully improving the experience for developers even more.
I may have exaggerated things a bit but I genuinely think this library has evolved a lot over time and I wish for the best.
Please feel free to correct me on any mistakes or incorrect assumptions that I may have made as I'm not a perfect being.

As for the unfortunate souls that have decided to read this completely: Thank you.
And I have a question for you; Have you had a similar experience with this library? I can't be the only one, right?
I would like to hear your stories!

Best regards, Folly :3

**Comment by martindevans** [comment]
Thank you for posting this! I experienced exactly the same initial pain when I first looked at LlamaSharp about a year ago. That was actually what motivated me to join the project, contribute improvements and eventually develop the `BatchedExecutor`! I've always been a bit worried it's too "low level" and exposes too much, so it's really good to get feedback like this!

Now that you've used the system, how would you improve it?

My long term plan has always been to build higher level executors over the batched executor (probably wrapping a single conversation object). Does that seem like it would improve things to you?

**Reply by follyfoxe** [reply]
You're welcome!!
I believe making higher level executors wrapping a single conversation is the next logical step, in fact, that's how I'm currently doing things; I have my own wrapper class that handles chat messages and such via the executor.
I also believe working directly with tokens should be encouraged because ideally, you don't want the user inserting special tokens and breaking everything; But you also don't want to disable them entirely.
In an ideal world, high level executors would take care of prompt formatting in this way.

About improvements, something I'd like is a teeny tiny bit more documentation/examples, specially some clarification regarding the processing of tokens; What if I wanted to "pre-process" tokens? That is, to feed the tokens into the context but not generate any new tokens. Should I just call `BatchedExecutor.Infer` until `Conversation.RequiresInference` is `false`, perhaps like the [llava example](https://github.com/SciSharp/LLamaSharp/blob/master/LLama.Examples/Examples/BatchedExecutorLLava.cs)? Or is there a better, more efficient way?
I have more or less an idea on how things work behind the scenes, but have knowleadge gaps and can't find much information about it.
Also I'd love to be able to get text embeddings without having to create a new context with `LLamaEmbedder` although it isn't much of a problem and don't know if that would be possible.

**Reply by martindevans** [reply]
> In an ideal world, high level executors would take care of prompt formatting in this way.

Definitely agree. In fact I started some work laying the foundations for this with the [`LlamaTemplate`](https://github.com/SciSharp/LLamaSharp/blob/LLama/LLamaTemplate.cs) class.

At the moment that's not ideal because it outputs all the text (template text, plus message content in one blob) but as I understand it this should be tokenized separately - template text should use special=true (so templates can output special tokens) and messages should use special=false (so users can't just type special tokens and override the system prompt). This is a design constraint inherited from llama.cpp and it'll probably need to be fixed upstream before we can make progress here.

> About improvements, something I'd like is a teeny tiny bit more documentation/examples

I've been putting together a few examples of random things when I get the time, all using `BatchedExecutor` to try and expand examples of what can be done with it. e.g. forking many times, implementing beam search, batched execution of a BoolQ evaluation etc. Any specific ideas for new examples I could put together?

> What if I wanted to "pre-process" tokens? That is, to feed the tokens into the context but not generate any new tokens. Should I just call BatchedExecutor.Infer until Conversation.RequiresInference is false, perhaps like the [llava example](https://github.com/SciSharp/LLamaSharp/blob/master/LLama.Examples/Examples/BatchedExecutorLLava.cs)? Or is there a better, more efficient way?

To answer this one directly - what you suggest is fine.

In more detail: llama.cpp can process tokens in batches, processing N tokens together in a batch is much cheaper than processing N tokens 1 by 1. A batch can contain a maximum number of tokens (e.g. 512 is a normal size) prompting any number of sequences (aka conversations). When a batch is run the logits (outputs from the LLM) are made available in a buffer, when you call `Sample` on a conversation it's really giving you a pointer into that buffer. This is why you must call `Sample` only after the correct batch has been run - that buffer will have been re-used for other batches if you read from it later, so it'll be meaningless! `BatchedExecutor` is checking this, and that's what the `RequiresInference` flag indicates.

`BatchedExecutor` internally keeps a queue of unprocessed work (a queue of batches). Once the last batch in the queue fills up, another one is added to the queue and it'll start adding work to that batch. Calling `Infer` simply takes the next batch from the queue and processes it.

To pre-process tokens you can simply call `Infer` until the queue is empty, or until the conversation indicates `RequiresSampling = false`. That will be submitting batches of work as quickly as possible, keeping the GPU as busy as possible.

Hopefully that all makes sense, if not please feel free to ask :)

> I'd love to be able to get text embeddings without having to create a new context

Embeddings are definitely an area that could do with a lot of work! If you're interested in contributing improvements there that'd be great!

The most obvious huge improvement that could be made is batched embeddings. At the moment we process them one by one, taking no advantage of batching - but exactly the same system that is used for the `BatchedExecutor` could also be used for embeddings. Potentially that could be a new type of object running within a `BatchedExecutor` (e.g. `EmbeddingsConversation` as a parallel to `Conversation`).

I think the reason the embedder creates a new context at the moment is because `Embeddings = true` must be set in the params, and I _think_ that used to prevent generating text (i.e. the entire context was in embeddings mode instead of text mode). However, I don't think that's the case any more.

In fact if you're interested in contributing, simply investigating the exact state of embeddings mode and whether or not it interferes with text generation and reporting back what you find would be a great start!

**Reply by follyfoxe** [reply]
> Definitely agree. In fact I started some work laying the foundations for this with the [LlamaTemplate](https://github.com/SciSharp/LLamaSharp/blob/LLama/LLamaTemplate.cs) class.

I did look at that class a while ago but decided not to use it since it doesn't allow for generating user responses for example (Which I like to experiment with sometimes)
But I remember looking at the code and seeing it's a llama.cpp limitation, so it doesn't matter that much. Moreover, not being able to do that is good enough for most cases. I'm probably the exception.

> Any specific ideas for new examples I could put together?

Perhaps a plain minimal example for general inference, but apart from that nothing comes to mind!

> When a batch is run the logits (outputs from the LLM) are made available in a buffer, when you call Sample on a conversation it's really giving you a pointer into that buffer. This is why you must call Sample only after the correct batch has been run - that buffer will have been re-used for other batches if you read from it later, so it'll be meaningless!

This makes sense and I was able to figure it out just by looking at the usage, quite intuitive!

> BatchedExecutor internally keeps a queue of unprocessed work (a queue of batches). Once the last batch in the queue fills up, another one is added to the queue and it'll start adding work to that batch. Calling Infer simply takes the next batch from the queue and processes it.

> To pre-process tokens you can simply call Infer until the queue is empty, or until the conversation indicates RequiresSampling = false. That will be submitting batches of work as quickly as possible, keeping the GPU as busy as possible.

What if I were to for example, `Prompt` the conversation with lots of tokens, in such a way that it has to create multiple batches. Doesn't that mean that I would have to call `Infer` multiple times?
Examples currently call this once, perhaps assuming the prompt will never get that large. This led me to initially think calling `Infer` would process all the tokens.
If that's the case, and I'm not mistaken, then this detail should be added somewhere, maybe on the method's documentation?

> I think the reason the embedder creates a new context at the moment is because Embeddings = true must be set in the params, and I think that used to prevent generating text (i.e. the entire context was in embeddings mode instead of text mode). However, I don't think that's the case any more.

> In fact if you're interested in contributing, simply investigating the exact state of embeddings mode and whether or not it interferes with text generation and reporting back what you find would be a great start!

I just did a quick test on the latest LLamaSharp release (0.13.0 as of writing) and can confirm, that setting `Embeddings = true` does not affect the ability to generate logits. My app works perfectly fine and generates text.
Also, in the property's documentation it says the following `If true, extract embeddings (together with logits).` ([`IContextParams`](https://github.com/SciSharp/LLamaSharp/blob/master/LLama/Abstractions/IContextParams.cs))

Unless llama.cpp says otherwise, I believe embeddings mode works as an optional addition :3

**Reply by martindevans** [reply]
> Doesn't that mean that I would have to call Infer multiple times?

Yes that's correct.

> Examples currently call this once, perhaps assuming the prompt will never get that large.

Yep, just laziness on my part when putting the examples together. I think I was trying to keep in simple, but if it's _too_ simple tat it mislead you then I might go through all the examples and fix that!

> I believe embeddings mode works as an optional addition

Thanks for checking that 👍

**Reply by martindevans** [reply]
Ah, I just dig some more digging into the embeddings thing and it looks like that's about to change. This code was added fairly recently into `llama_decode`:

```cpp
    // TODO: use a per-batch flag for logits presence instead
    const bool has_logits = !cparams.embeddings;
```

and this has been added into the header (I just noticed it while making a start on a new binary update):

```cpp
    // Set whether the model is in embeddings mode or not
    // If true, embeddings will be returned but logits will not
    LLAMA_API void llama_set_embeddings(struct llama_context * ctx, bool embeddings);
```

**Reply by follyfoxe** [reply]
> Yes that's correct.

Alright, going to take that into consideration going forward!

> Ah, I just dig some more digging into the embeddings thing and it looks like that's about to change.

Awww... that's a shame.
```cpp
// TODO: use a per-batch flag for logits presence instead
const bool has_logits = !cparams.embeddings;
const bool has_embd = cparams.embeddings && (cparams.pooling_type == LLAMA_POOLING_TYPE_NONE);
```
It's weird too, because I always thought text models like these *needed* the embeddings, because they didn't take tokens directly.
Hopefully it'll be possible in the future.

Anyways, thanks for everything!

**Reply by martindevans** [reply]
Thinking about this some more - it might be possible to make this all work within one executor.

At the moment the `BatchedExecutor` has a queue with 2 types of batch: tokens and embeddings (inputs). We could add a third type of batch, Embeddings (outputs), along with a new type of conversation that only allows extracting output embeddings. The context could then be switched in and out of embedding mode (using `llama_set_embeddings`) per batch as necessary!

**Comment by SerialKicked** [comment]
Hey,

I'm a bit confused about the use case for the BatchedExecutor. I get the concept for the others, and I personally use the StatelessExecutor, handling prompt formatting, token count, logging, and instruct formats myself (my system prompt changes dynamically, making stateful executors a bit pointless for me). 

From what I gathered from the doc (btw, thank you for having a coherent doc. It's beyond rare in the AI field, sadly), it seems that it's used to process multiple prompts at the same time? Unless it's designed to run on multiple graphic cards, I don't really get the benefit. And if it's sequentially, instead of parallel, I'm literally lost :D 

I get it's a lower level way to access the model, but I don't understand the use case. I'm just curious as I'm fundamentally unable to learn something without a concrete use case / example, no matter the amount of documentation available. I maybe could see it for when i'm running a small embedding model to feed a ton of message history and arbitrary text into a Vector DB, at vaguely higher speed than when using Stateless in a loop, but if feel like I'm missing something important.

**Reply by martindevans** [reply]
The `BatchedExecutor` (despite it's name) is not _just_ for batched execution. It is the only executor we have that can do batching (hence the name) but it was built as a new foundation for all future executors. My intention is for future high level executors to be built as wrappers around a single `Conversation` object. Basically it's a lower level, more flexible executor that assumes less and exposes more capabilities (while try to ensure you can't mis-use it, by checking and throwing exceptions).

> concrete use case / example

The advantages are:
1) Less stateful/confusing
2) Batching lots of conversations is faster and more memory efficient
3) More control over KV cache

Let's go through these one by one.

## Less State

Personally when I first started using LLamaSharp I found the high level executors (in particular instruct and interact) _extremely_ confusing, by the sound of it so did @follyfoxe. The main confusion for me was the amount of state they hold - they're try to do too many things in one class and it's just very difficult to keep track of what state it's in.

Conversely `BatchedExecutor` tries to be absolutely minimal. Higher level things can be built over it, keeping the core simple, and if someone wants something high level that we don't provide they can build their own wrapper around `BatchedExecutor`. That way they don't have to drop all the way down to the low level llama.cpp API (which is very easy to mis-use).

## Batching lots of conversations is faster

This ties into what you said:

> Unless it's designed to run on multiple graphic cards, I don't really get the benefit.
> feel like I'm missing something important.

What you're missing is that executing a batch of e.g. 10 sequences at once is faster then executing e.g. 10 sequences one by one. This is because the main bottleneck is memory bandwidth not compute. For example for a single token you need to (times are made up, for illustration):

1) Load layer 1 (10ms)
2) Process single token with layer 1 (1ms)
3) Load layer 2 (10ms)
4) Process single token with layer 2 (1ms)
5) etc

Wheras say you're processing 10 sequences (each adding 1 token), you do:
1) Load layer 1 (10ms, same as before)
2) Process 10 tokens with layer 1 (10ms, 10x as long as before)
3) Load layer 2 (10ms, same as before)
4) Process 10 tokens with layer 2 (10ms, 10x as long as before)
5) etc

You get to do a lot more work at each layer, so you waste less time loading things from memory overall. In this example (**with totally made up times**) we first did 22ms for 1 token, then we did 40ms for 10 tokens which is 4ms/token or 5.5x better throughput!

## Batching lots of conversations is more memory efficient

One of the things you can do with a `Conversation` object is `Fork()`. This allows you to create a copy of an existing `Conversation` and **all of the memory up to that point is shared between the two sequences**, in fact in terms of memory usage forking is basically free! This can be used for example to evaluate a long document (just once) and then fork that conversation to ask many individual questions (like a stateless executor per question, but skipping the prompt evaluation cost every time and also saving a lot of memory if there are several parallel multi-turn conversations with the same document prompt).

> More control over KV cache

Let's say you were using the `StatelessExecutor` and you fed in enough of a prompt that you ran out of context space - what would happen? The `StatelessExecutor` is currently hardcoded to shift the input prompt left, keeping some amout at the start. I hope that's the behaviour that you want because it's not optional and you can't use anything else (e.g. context self-extend).

A `Conversation` exposes more control over the KV cache. If you _want_ that shifting behaviour you can just call `ShiftLeft()`. However, if you want something else you can add your own low level KV-cache manipulations, note how `Rewind` and `ShiftLeft` are implemented as extension methods ([here](https://github.com/SciSharp/LLamaSharp/blob/master/LLama/Batched/ConversationExtensions.cs)) - you could have implemented those in your own project, they're not using any secret internal APIs. Even context self extension could probably be implemented as an extension method with 5 or so lines of code.

Hopefully that helps, feel free to ask for more detail if you don't understand anything!

**Reply by SerialKicked** [reply]
Wow, thank you!

That, actually, was a great explanation. You probably should copy-paste that in your documentation. Not only, I get it now, but it also answers several questions and issues I had while testing/using the other executors.

> Less State

Oh yes, I do agree. Entirely. I found the stateful executors, at best superfluous, and at worse, counter-productive. The instruct one, for instance doesn't really do instruct format handling. I know about the text transformation thingy you can override, and that you can add prefixes to the input. But that ain't it.

My attempts at using LLamaSharp are kinda similar to OP, except I went in the order they were shown in the doc. So I wasted my time on the 2 stateful ones, gave up, considered becoming a front-end for OAI-protocol compatible back-ends instead, gave it another go at 0.13, but instead of going for the batched one, I went for stateless instead.

Funny, you mention the low level API, after that initial attempt, I was *that close* to just port the DLL myself and try my chance with it. I probably would have wasted a LOT of time doing that.

> Batching lots of conversations is faster

Gotcha. I get the technical reason why it would be beneficial in some situations. 

> One of the things you can do with a Conversation object is Fork(). This allows you to create a copy of an existing Conversation and all of the memory up to that point is shared between the two sequences, in fact in terms of memory usage forking is basically free! This can be used for example to evaluate a long document (just once) and then fork that conversation to ask many individual questions [...]

Now, we're talking! That, I have many many use cases for. And, yes, it makes a lot of sense now.

> Let's say you were using the StatelessExecutor and you fed in enough of a prompt that you ran out of context space - what would happen? The StatelessExecutor is currently hard-coded to shift the input prompt left, keeping some amount at the start. I hope that's the behaviour that you want because it's not optional and you can't use anything else (e.g. context self-extend).

I didn't know that about StatelessExecutor, I assumed it'd throw an exception at me ("your prompt + response length > context window" type of thing), or feed it nonetheless to the model, letting it break on its own, but do not that. That explains so many of the issues I've encountered now. Arbitrarily shifting an instruct-formatted prompt can completely fuck up the model's output. While I know that I'm not supposed to send more than the context size, it'd be a lot more useful to get an exception when I fucked up my count (which likely happened several times during my development) than the issue being pulled under the rug by the executor.

It really should be indicated somewhere in documentation how the executors individually handle prompt>context. It's kinda crucial information. I *guessed* the stateful executors handled it as they have some way to do it correctly. I wasn't expecting that of the stateless one. 

Welp, time to rewrite quite a bit of my program, and switch to batched. 

--

Anyway, thank you again for the extremely detailed clarifications! It's really appreciated!

Cheers!

**Reply by webitube** [reply]
@martindevans I've been experimenting with the BatchedExecutor recently and have gotten incredible performance with it when running many conversations.

But, things seem to fall apart when I try to use a Grammar with it. It seems like the Grammar works with the StatelessExecutor but not BatchedExecutor. Is constrained output not supported on the BatchedExecutor? Or maybe, I'm doing it wrong?

**Reply by martindevans** [reply]
Grammars are something applied to the logits after the executor has run, so it shouldn't matter which executor you're using. What error are you seeing?

**Reply by martindevans** [reply]
Btw the BoolQ benchmark example uses `BatchedExecutor` (to answer lots of questions at once) and `Grammar` (to constrain the output to just yes/no). Might serve as a good example: https://github.com/SciSharp/LLamaSharp/blob/master/LLama.Examples/Examples/BatchedExecutorBoolQ.cs

**Reply by webitube** [reply]
@martindevans 

I created a very simple grammar to return an answer to a question in the following form:
`   {"answer":"answer text"}`

grammar: 
```
{
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "type": "object",
    "required": [
        "answer"
    ],
    "properties": {
        "answer": {
            "type": "string"
        }
    }
}
```
The response I get back is just a long string of, "{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{{".

I finally got my code working. I'm still distilling the changes to just the necessary ones, but, after reading BoolQ and BatchedExecutorGuidance (on which my work is based), I think I know what the problem probably is.

I think the problem is in BaseSamplingPipeline.Sample(). BatchedExecutorGuidance (DefaultSamplingPipeline :  BaseSamplingPipeline.Sample()) which never calls Accept on the output token whereas BoolQ does this manually in Prompt().

Specifically, I suspect the problem is in the last line of BaseSamplingPipeline.Sample():
`   return ProcessTokenDataArray(ctx, candidates, lastTokens);`

In my version, I changed this to:
```
   var token = ProcessTokenDataArray(ctx, candidates, lastTokens);
   Accept(ctx, token);
   return token;
```
   
Note that my version has many other changes, but my best guess right now is not Accepting the token is the issue.

I'll report back after I confirm my findings.

**Reply by martindevans** [reply]
Ah yes without a call to `Accept` the grammar doesn't know that a token has been picked, so it keeps trying to pick the first one. `Accept` is meant to be called by whatever is using the pipeline, rather than the pipeline itself. For example you can see the `StatelessExecutor` doing tha [here](https://github.com/SciSharp/LLamaSharp/blob/master/LLama/LLamaStatelessExecutor.cs#L111-L112).

**Reply by webitube** [reply]
Thanks @martindevans! I understand now.

To sum up looking at the original `BatchedExecutorGuidance`, the problem I had with adding Grammar support is twofold. First, the token needed to be accepted after `Sample()`. I added a separate `pipeline.Accept()` right after the Sample() on the example application side. Second, the example pipeline used, `GuidedSampler`, overrides `Accept()` to do nothing. Once this override is removed, it works fine.

Should `BatchedExecutorGuidance` be updated to do the `Accept()` and remove the `GuidedSampler.Accept()` override? It works fine as-is, but it could be confusing to future users/readers if they want to enable constrained output.

**Reply by martindevans** [reply]
Yeah I'm not sure why `Accept` is being overridden there. I think possibly it used to be an abstract method (so I just implemented it to do nothing, since it's not relevant to guidance), but now it's just a bit confusing! If you'd like to submit a PR removing that I'll be happy to review and merge that :)

**Reply by webitube** [reply]
@martindevans I created a PR for this change: https://github.com/SciSharp/LLamaSharp/pull/904
